Update after PR27394 to re-add bootstrap ignition config file

bobfuru opened the pull request:
**Edit (9/22/21):** This PR should address recent issues reported in https://bugzilla.redhat.com/show_bug.cgi?id=1906890.

**PREVIEW BUILD LINK:** https://deploy-preview-28106--osdocs.netlify.app/openshift-enterprise/latest/installing/installing_vmc/installing-restricted-networks-vmc-user-infra.html#installation-vsphere-machines_installing-restricted-networks-vmc-user-infra

In https://github.com/openshift/openshift-docs/pull/27394, we removed a couple of steps about uploading the `bootstrap.ign` file to the http server based on dev comments that this was no longer necessary [1],[2]. But this seems to only apply in install-provisioned and not user-provisioned installs, so this probably should not have changed. 

In this PR, I'm adding the steps back in, slightly reworded for clarity and to account for @dav1x 's comment that these steps aren't necessary if you store the bootstrap ignition in the VM's extra config properties [2].

~~Also, this PR updates the instructions to use `Ignition spec 3.1.0` in OCP 4.6 and up by replacing `append` with `merge` in the command examples [3].F~~

[1] https://github.com/openshift/openshift-docs/pull/27394#discussion_r526327784
[2] https://github.com/openshift/openshift-docs/pull/27394#discussion_r537834967
[3] https://github.com/RedHatOfficial/ocp4-vsphere-upi-automation/issues/49

~~Once this is approved and merged, separate PRs should be created against enterprise-4.5 and enterprise-4.4 to use similar wording (although without the `merge` wording update for `Ignition spec 3.1.0`)~~. 💫 

**Edit:** The `ignition.config.merge` portion of this PR was removed when the step to create a secondary Ignition config file was removed [here](https://github.com/openshift/openshift-docs/pull/27394/files#diff-0180108d51ee983e7a96405a2e19b7bde93377496523b317c727d1f89a6f2f24L30-L80).

bobfuru commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -17,6 +17,38 @@ Before you install a cluster that contains user-provisioned infrastructure on VM
+
+. Upload the file to your HTTP server.  Note the URL of this file.
+
+. Save the following secondary Ignition config file for your bootstrap node to your computer as `<installation_directory>/merge-bootstrap.ign`.
```
Calling out here that I changed the filename from what we have in 4.5 docs to reflect `Ignition spec 3.1.0`: 
s/`<installation_directory>/_**append**_-bootstrap.ign`/`<installation_directory>/**_merge_**-bootstrap.ign`

bobfuru commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -17,6 +17,38 @@ Before you install a cluster that contains user-provisioned infrastructure on VM
+{
+  "ignition": {
+    "config": {
+      "merge": [
```
Calling out here that I changed the filename from what we have in 4.5 docs to reflect `Ignition spec 3.1.0`: 
`"append":  [` --> `"merge":  [`

bobfuru commented:
Hi @jcpowermac, @patrickdillon - Could you PTAL to see if these changes look good? 

The previous edits we made in PR27394 were to the UPI docs but it seems that it would only be true for IPI that you no longer need to host bootstrap Ignition on the webserver. 

Thanks!

bobfuru commented:
Cc: @rhatbob

jcpowermac commented:
> Hi @jcpowermac, @patrickdillon - Could you PTAL to see if these changes look good?
> 
> The previous edits we made in PR27394 were to the UPI docs but it seems that it would only be true for IPI that you no longer need to host bootstrap Ignition on the webserver.
> 
> Thanks!

IPI never needed the bootstrap webserver. UPI needed just because originally the terraform used vApp instead of extra config. Might as well make it easier on customers to use extra config vs setting up a webserver.

bobfuru commented:
> IPI never needed the bootstrap webserver. UPI needed just because originally the terraform used vApp instead of extra config. Might as well make it easier on customers to use extra config vs setting up a webserver.

Thanks, @jcpowermac - I've removed the webserver instructions again. 😛  Not sure if the example I added for extra config properties looks right:

```
resource "vsphere_virtual_machine" "vm" {
  for_each = var.hostnames_ip_addresses
  extra_config = {
    "<installation_directory>/bootstrap.ign"   <1>
  }
}
<1> Specify the installation directory of the bootstrap Ignition config file that you hosted.
```
Or should I remove the code block out and just leave, "Store the Ignition config file in the extra configuration properties of the virtual machine (VM)" as step 2. WDYT?

bobfuru commented:
@jcpowermac bump - wondering if the example I added for extra config properties looks good so I can merge.

bobfuru commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -12,11 +12,26 @@ Before you install a cluster that contains user-provisioned infrastructure on VM
 .Prerequisites
 
 * Obtain the Ignition config files for your cluster.
-* Have access to an HTTP server that you can access from your computer and that the machines that you create can access.
```
No longer hosting bootstrap Ignition config on webserver, so this prereq isn't necessary.

bobfuru commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -34,7 +49,7 @@ $ base64 -w0 <installation_directory>/worker.ign > <installation_directory>/work
 +
 [source,terminal]
 ----
-$ base64 -w0 <installation_directory>/append-bootstrap.ign > <installation_directory>/append-bootstrap.64
+$ base64 -w0 <installation_directory>/bootstrap.ign > <installation_directory>/bootstrap.64
```
I changed the reference to the custom Ignition config. (`s/append-bootstrap.ign/bootstrap.ign`) It was a remnant of the previous steps that we removed in PR27394.

bobfuru replied:
Updated to `merge-bootstrap.ign` in this PR because customer was reporting that `append-` was still needed but my understanding is that `append-bootstrap --> merge-bootstrap` starting in 4.6 due to update to the Ignition spec.

bobfuru commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -118,15 +133,18 @@ $ export IPCFG="ip=192.168.100.101::192.168.100.254:255.255.255.0:::none nameser
 **** `guestinfo.ignition.config.data`: Paste the contents of the base64-encoded Ignition config file for this machine type.
 **** `guestinfo.ignition.config.data.encoding`: Specify `base64`.
 **** `disk.EnableUUID`: Specify `TRUE`.
-*** Alternatively, prior to powering on the virtual machine, use vApp properties to:
-**** Navigate to a virtual machine from the vCenter Server inventory.
-**** On the *Configure* tab, expand *Settings* and select *vApp options.*
-**** Scroll down and under *Properties*, apply the configurations that you just edited.
```
Removed the "alternatively" step per @jcpowermac.

bobfuru commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -118,15 +133,18 @@ $ export IPCFG="ip=192.168.100.101::192.168.100.254:255.255.255.0:::none nameser
 $ govc vm.change -vm "<vm_name>" -e "guestinfo.afterburn.initrd.network-kargs=${IPCFG}"
 ----
 +
+... Set the extra config properties of the control plane, compute, and bootstrap Ignition config files before booting a VM from an OVA in vSphere. For example, to change the bootstrap Ignition config:
++
+[source,terminal]
+----
+$ govc vm.change -vm "<vm_name>" -e "guestinfo.bootstrap.ign=${IPCFG}"
+----
++
```
From Slack comment: 

> 5g where it describes afterburn with govc commands. We could do that for ignition config as well

I'm not sure if this is suggesting to place this in the first steps where the user is locating and converting Ignition configs, or alongside this step (5g) with afterburn. I've gone with the latter, though I don't understand what the IPCFG variable is doing and if it should apply. 

@jcpowermac Could you specify where you think this step belongs and what command the user should run? Is it also optional?

jcpowermac replied:
@bobfuru this is kinda what I was thinking:
```
govc vm.change -vm "<vm_name>" -e "guestinfo.ignition.config.data.encoding=base64"
govc vm.change -vm "<vm_name>" -e "guestinfo.ignition.config.data=$(cat bootstrap.ign)"
```
Would need to be done to all 7 virtual machines though.

bobfuru replied:
IIUC, these steps are better suited for the beginning of the procedure. Basically, they should find the bootstrap, master, and worker Ignition configs, convert to Base64, and store in the ExtraConfig of the VM using your command examples. 

Assuming that to be the case, I relocated these to the top (steps 1-3).

jcpowermac commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -12,11 +12,26 @@ Before you install a cluster that contains user-provisioned infrastructure on VM
 
+. Locate the bootstrap Ignition config file `<installation_directory>/bootstrap.ign` that the installation program created.
+
+. Store the Ignition config file in the extra configuration properties of the virtual machine (VM). For example:
```
@bobfuru I am unsure this is useful to add a snip of terraform. It can't be executed as-is and since terraform isn't listed might confuse the reader what we are trying convey.

bobfuru replied:
Ack. Removed the terraform sample.

jcpowermac commented:
@patrickdillon can you please review to make sure you agree with my comments

bobfuru commented:
@jcpowermac and @patrickdillon - Updates made based on Joe's command examples. To me, this makes more sense now and resolves https://github.com/openshift/openshift-docs/issues/28200.

Please review and share your thoughts. Thanks.

bobfuru commented:
@jcpowermac and @patrickdillon - BUMP, PTAL to be sure this looks good in how it describes using ExtraConfig rather than webserver to store Ignition configs.

I'd like to get this merged so that the docs are correct. Thanks!

Preview link (internal):
http://file.rdu.redhat.com/bfuru/012021/restore-vsphere-ipi-steps/installing/installing_vsphere/installing-vsphere.html#installation-vsphere-machines_installing-vsphere

madchr1st commented:
@jcpowermac @bobfuru @patrickdillon - I have tried setting the ignition config with the govc binary as described above and got the error stating the argument was too long.
IMHO this is a limitation of the extra parameters on vCenter side and therefore we still need the append-bootrstrap option with a webserver.
Am  I doing something wrong or have you not tested this solution?

jcpowermac commented:
@madchr1st it has been tested though not with govc. I am not completely "sold" on govc over powercli; seen various issues.
https://gist.github.com/jcpowermac/501a7db7e875acd2efb83bbb9250a02f

@rvanderp3 had a similar issue with that I think it was caused by a bash limitation. 

There is no webserver or append boostrap in vSphere IPI it _only_ uses extra config.

bobfuru commented:
I'm unable to access a vSphere cluster to test this procedure and cannot verify that there are problems, as @madchr1st reports. Can someone recommend next steps? @jcpowermac @patrickdillon 
Cc: @miabbott @mike-nguyen

oarribas commented:
@bobfuru , the information about the change of `append` with `merge` [mentioned in the first comment](https://github.com/openshift/openshift-docs/pull/28106#issue-539767295) is now missing in the commit. Maybe because the files are created with `merge` by the installer. Can you clarify this?

bobfuru commented:
@oarribas - IIRC, the reason the append/merge is now missing is because we removed a step for saving a secondary Ignition config file that contained `ignition.config.append/merge` until 4.6 as initially suggested in this comment: https://github.com/openshift/openshift-docs/pull/27394#discussion_r526327784
[Here](https://github.com/openshift/openshift-docs/pull/27394/files#diff-0180108d51ee983e7a96405a2e19b7bde93377496523b317c727d1f89a6f2f24L30-L80) is where that step was removed in the PR.
Does that help?

bobfuru commented:
@jcpowermac - Any new thoughts on whether this PR could get added back in the docs? Not sure where we'd landed with using `govc` in this procedure.
Cc: @miabbott @mike-nguyen

miabbott commented:
Sadly, RHCOS QE doesn't have a dedicated vSphere setup to validate these changes.  I'd defer to @jcpowermac and @rvanderp3 for the more reliable way of interacting with the vSphere cluster.

mike-nguyen commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -16,24 +16,49 @@ Before you install a cluster that contains user-provisioned infrastructure on VM
 [source,terminal]
 ----
-$ base64 -w0 <installation_directory>/master.ign > <installation_directory>/master.64
+$ govc vm.change -vm "<vm_name>" -e "guestinfo.ignition.config.data=$(cat master.ign)"
```
Followup to Micah's comment https://github.com/openshift/openshift-docs/pull/28106/files#r635516210, it would look like this for the `base64` conversion

```suggestion
$ govc vm.change -vm "<vm_name>" -e "guestinfo.ignition.config.data=$(base64 -w0 master.ign)"
```

bobfuru replied:
Thanks, @mike-nguyen - you read my mind! I've resolved the 3 places where this occurs.

bobfuru commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -16,24 +16,49 @@ Before you install a cluster that contains user-provisioned infrastructure on VM
++
+. Convert the control plane, or master, Ignition config files to Base64 encoding and store them in the extra configuration properties of the control plane VMs before booting.
++
+.. To convert to Base64 and change the control plane, or master, Ignition config files, enter the following commands for each `master.ign` file:
```
That makes sense, thanks. Resolved 3 occurrences of this based on Mike's suggestion.

bobfuru commented:
Updated proper Base64-conversion based on RHCOS QE comments.
**New preview build (internal):** http://file.rdu.redhat.com/bfuru/0521/restore-vsphere-ipi-steps/installing/installing_vmc/installing-restricted-networks-vmc-user-infra.html#installation-vsphere-machines_installing-restricted-networks-vmc-user-infra

@jcpowermac PTAL for approval, thanks.

bobfuru commented:
@jcpowermac or @rvanderp3 BUMP for SME ack, PTAL. Thanks!

jcpowermac commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -19,24 +19,49 @@ Before you install a cluster that contains user-provisioned infrastructure on VM
++
+[source,terminal]
+----
+$ govc vm.change -vm "<vm_name>" -e "guestinfo.ignition.config.data=$(base64 -w0 bootstrap.ign)"
```
This doesn't work because of bash length limitations.
```
[jcallen@fedora vsphere-upi]$  govc vm.change -vm "jcallen-registry" -e "guestinfo.ignition.config.data=$(base64 -w0 bootstrap.ign)"
bash: /home/jcallen/bin/govc: Argument list too long
[jcallen@fedora vsphere-upi]$
```
PowerShell (w/PowerCLI) does though
```
$vm | New-AdvancedSetting -name "guestinfo.ignition.config.data" -value $Ignition -confirm:$false -Force > $null
```
PowerCLI is probably more used by VMware administrators than govc. (I have no data to back this up, just that PowerCLI has been around for significantly longer period of time)

bobfuru replied:
Thanks, @jcpowermac. Is the PowerCLI command necessary to run separately for bootstrap, master, and worker nodes? (If so, how does it distinguish each?)

Or do they _not_ need step 1 and for step 2, do they only need to run the command once like below?

> ~2~1. Store the Ignition config file in the extra configuration properties of the virtual machine (VM) before booting by running the PowerCLI (`vm`) command:
> 
> `$ vm | New-AdvancedSetting -name "guestinfo.ignition.config.data" -value $Ignition -confirm:$false -Force > $null`

bobfuru commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -14,29 +14,25 @@ To install {product-title} on user-provisioned infrastructure on VMware vSphere,
 [source,terminal]
 ----
-$ base64 -w0 <installation_directory>/bootstrap.ign > <installation_directory>/bootstrap.64
+$ vm | New-AdvancedSetting \
+-name "guestinfo.ignition.config.data" \
+-value $Ignition -confirm:$false -Force > $null
```
Hi @jcpowermac - I've added the PowerCLI command here per your last comment. But I'm not clear as to whether this can be run once for all Ignition config files, or if it needs to be run individually for bootstrap, master, and worker Ignition config files? Could you review and let me know what's missing?

Here's the latest build preview: http://file.rdu.redhat.com/bfuru/0820/restore-vsphere-ipi-steps/installing/installing_vmc/installing-restricted-networks-vmc-user-infra.html#installation-vsphere-machines_installing-restricted-networks-vmc-user-infra

Thanks!

bobfuru commented:
Hi @mike-nguyen - Is it possible to review this updated procedure? I added some steps for converting Ignition configs to base64 using PowerCLI that I received from @jcpowermac, but not sure these will actually work or if we can support it in our documentation (Cc: @control-d).

I know RHCOS doesn't have a vSphere env for testing but it was mentioned to me that we can request access to VMC using [this Source link](https://source.redhat.com/groups/public/atomicopenshift/atomicopenshift_wiki/openshift_onboarding_checklist_for_vmware_cloud_on_aws_access). I was able to do so, and I can run PowerCLI, but I have been unsuccessful in deploying a vSphere cluster or in getting these `$vm` commands to run.

If this set of `vm` commands in step 2 can't work or be tested, I propose that we go back to using the previous documented method of uploading the Ignition config to your web server, missing since the OCP 4.5 doc [here](https://docs.openshift.com/container-platform/4.5/installing/installing_vsphere/installing-vsphere.html#installation-vsphere-machines_installing-vsphere). Thanks!

jinyunma commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -14,31 +14,56 @@ To install {product-title} on user-provisioned infrastructure on VMware vSphere,
+[source,terminal]
+----
+$ vm = New-VM -VM $Template -Name $name -ResourcePool $rp \
+  -Datastore $datastore -Location $Folder
```
Per my understanding,  the purpose is to use PowerCLI to set bootstrap based64 encoded ignition config to extra config `guestinfo.ignition.config.data.encoding`, other parts including cloning vm from template, customized hardware etc still need to do through vSphere client, or you plan to change by using PowerCLI also for those parts?

bobfuru replied:
This is my understanding as well. PowerCLI is just used in this step to convert Ignition config files to Base64 encoding. I don't plan to change other parts of this doc to use PowerCLI. @jcpowermac can you confirm?

jinyunma commented:
I tested that following two ways both works:
1.  cope the base64 ecnoded bootstrap ignition config and paste to extra config "guestinfo.ignition.config.data"
2.  use PowerCLI to set extra config "guestinfo.ignition.config.data"
`$vm = Get-VM -name 'control-plane-2' -Location 'jima1109a-mcgds'`
`$vm | New-AdvancedSetting -name "guestinfo.ignition.config.data" -value $(cat bootstrap.64) -confirm:$false -Force`

bobfuru commented:
Thanks for your feedback, @jinyunma. I haven't been able to get consensus on a tested and verifiable procedure that would replace the web server with either govc or PowerCLI.

I've therefore opted to re-add the section on using an HTTP server for the alternate bootstrap Ignition config file, as was originally the goal with this PR. 

@jcpowermac @jinyunma Could you PTAL for SME and QE review?

jinyunma commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -14,13 +14,52 @@ To install {product-title} on user-provisioned infrastructure on VMware vSphere,
+      ]
+    },
+    "timeouts": {},
+    "version": "2.2.0"
```
Different releases support the different ignition version, need to be updated based on release version.

bobfuru replied:
Fixed. Thanks for the catch!

jinyunma commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -129,13 +168,9 @@ $ govc vm.change -vm "<vm_name>" -e "guestinfo.afterburn.initrd.network-kargs=${
 *** Optional: In the event of cluster performance issues, from the *Latency Sensitivity* list, select *High*.
 *** Click *Edit Configuration*, and on the *Configuration Parameters* window, click *Add Configuration Params*. Define the following parameter names and values:
-**** `guestinfo.ignition.config.data`: Paste the contents of the base64-encoded Ignition config file for this machine type. Note for the bootstrap node, the Ignition config file must be provided in `guestinfo.ignition.config.data` in the *Configuration Parameters* window. This is due to a restriction in the maximum size of data that can be provided in a vApp property.
+**** `guestinfo.ignition.config.data`: Locate the base-64 encoded files that you created previously in this procedure, and paste the contents of the base64-encoded Ignition config file for this machine type. Note that for the bootstrap node, the Ignition config file must be provided in `guestinfo.ignition.config.data` in the *Configuration Parameters* window. This is due to a restriction in the m ...
```
Since bootstrap ignition config file has already been changed to be provided by `merge-bootstrap.64`, do we still need to describe it here? Or merge-bootstrap.64 is just an option for user who could not copy/paste bootstrap based64-encoded ignition config directly?

bobfuru replied:
Good point. Seems we don't need to describe the bootstrap Ignition config note here, so I've removed it.

bobfuru commented:
@jinyunma I've addressed your latest comments, let me know if this looks good from QE perspective.

jinyunma commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -14,13 +14,52 @@ To install {product-title} on user-provisioned infrastructure on VMware vSphere,
-
+. Upload the bootstrap Ignition config file, which is named `<installation_directory>/bootstrap.ign`, that the installation program created to your HTTP server. Note the URL of this file.
++
+You must host the bootstrap Ignition config file because it is too large to fit in a vApp property.
```
Since you have already removed "alternatively" step by using vApp, not sure if we still need to mention vApp here?

bobfuru replied:
Yes, that makes sense. This statement shouldn't be necessary without the alternative step of configuring vApp properties. Removed.

jinyunma commented:
@bobfuru other parts looks good to me except above minor comments.

bobfuru commented:
Addressed your feedback, @jinyunma - please move the BZ to verified if all looks good to you now. Thanks!

jcpowermac commented:
/lgtm

jinyunma commented:
/lgtm

kalexand-rh commented on modules/installation-vsphere-machines.adoc:
```diff
@@ -14,13 +14,49 @@ To install {product-title} on user-provisioned infrastructure on VMware vSphere,
-* Create a link:https://docs.vmware.com/en/VMware-vSphere/6.0/com.vmware.vsphere.vcenterhost.doc/GUID-B1018F28-3F14-4DFE-9B4B-F48BBDB72C10.html[vSphere cluster].
+* You have obtained the Ignition config files for your cluster.
+* You have access to an HTTP server that you can access from your computer and that the machines that you create can access.
+* You have created a link:https://docs.vmware.com/en/VMware-vSphere/6.0/com.vmware.vsphere.vcenterhost.doc/GUID-B1018F28-3F14-4DFE-9B4B-F48BBDB72C10.html[vSphere cluster].
```
Is this the right link for VMC, too?

bobfuru replied:
Seems so. It's a child topic of the [About VMC](https://docs.vmware.com/en/VMware-vSphere/6.0/com.vmware.vsphere.vcenterhost.doc/GUID-3B5AF2B1-C534-4426-B97A-D14019A8010F.html) doc. It is the same external link we've had for each OCP 4.x version, so I'm inclined to keep it here.

kalexand-rh reviewed (commented):
I have one question to consider, but this otherwise looks good.

bobfuru commented:
/cherrypick enterprise-4.6

bobfuru commented:
/cherrypick enterprise-4.7

bobfuru commented:
/cherrypick enterprise-4.8

bobfuru commented:
/cherrypick enterprise-4.9

bobfuru commented:
/cherrypick enterprise-4.10

bobfuru commented:
/cherrypick enterprise-4.10

bobfuru commented:
/cherrypick enterprise-4.6
